---
title: "Simpósio de Biologia Molecular"
subtitle: "Pipelines versáteis e escalonáveis para análises genômicas bacterianas completas"  
author: 
  - "Felipe M. Almeida"
  - "Departamento de Biologia Celular, Universidade de Brasília"
# date: '`r Sys.Date()`'
date: 'November 2021'
output:
  xaringan::moon_reader:
    includes:
      after_body: insert-logo.html
    css: [xaringan-themer.css, custom.css]
    nature:
      slideNumberFormat: "%current%"
      highlightStyle: github
      highlightLines: true
      ratio: 16:9
      countIncrementalSlides: true
---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(
  fig.width=9, fig.height=3.5, fig.retina=3,
  out.width = "100%",
  cache = FALSE,
  echo = TRUE,
  message = FALSE, 
  warning = FALSE,
  hiline = TRUE
)
library(knitr)
library(icons)
# source: https://github.com/gadenbuie/xaringanExtra
#xaringanExtra::use_animate_css()
#xaringanExtra::use_animate_all("slide_down")
#xaringanExtra::use_scribble()
xaringanExtra::use_search(show_icon = TRUE)
xaringanExtra::use_tachyons()
xaringanExtra::use_webcam()
xaringanExtra::use_freezeframe()
xaringanExtra::use_tile_view()
htmltools::tagList(
  xaringanExtra::use_clipboard(
    button_text = "<i class=\"fa fa-clipboard\"></i>",
    success_text = "<i class=\"fa fa-check\" style=\"color: #90BE6D\"></i>",
    error_text = "<i class=\"fa fa-times-circle\" style=\"color: #F94144\"></i>"
  ),
  rmarkdown::html_dependency_font_awesome()
)
xaringanExtra::use_fit_screen()
xaringanExtra::use_extra_styles(
  hover_code_line = TRUE,         #<<
  mute_unhighlighted_code = TRUE  #<<
)
```

```{r xaringan-themer, include=FALSE, warning=FALSE}
library(xaringanthemer)
style_duo_accent(
  primary_color = "#a79ebf",
  secondary_color = "#906068",
  link_color = "darkblue",
  inverse_header_color = "#FFFFFF",
  header_font_google = google_font("Josefin Sans"),
  text_font_google   = google_font("Montserrat", "300", "300i"),
  code_font_google   = google_font("Fira Mono")
)
# colors
  # primary_color = light_yellow,
  # secondary_color = dark_yellow,
  # header_color = gray,
  # text_color = gray,
  # code_inline_color = colorspace::lighten(gray),
  # text_bold_color = colorspace::lighten(gray),
  # link_color = blue,
  # title_slide_text_color = blue,
```

# Contexto

.pull-left[
* Sequenciamento de próxima geração (NGS) vem sendo utilizado com sucesso no contexto de pesquisas e acompanhamento de doenças infecciosas

    + Sequenciamento de genoma completo de patógenos
    
        + Alta resolução do conteúdo genético
        + Investigação do conteúdo de virulência, resistência, elementos móveis, etc.
        
    + Identificação de origens de epidemias
        + rastreamento de transmissões
        + Investigação da dinâmica

* Nova área: **Vigilância genômica**
]

???

* NGS data is being successfully used for the study and control of infectious diseases which is giving rise to a relatively new field called *Genomic surveillance*.

* The pillars of this filed is that the massive genome sequencing of pathogens enable the identification of outbreaks sources, track transmissions, investigate epidemic dynamics, investigate resistance and virulence traits, and more.

* All of that because of the high resolution genetic makeup of the pathogens provided by sequencing.

--
.pull-right[
.bg-washed-blue.b--dark-blue.ba.bw2.br3.shadow-5.ph4.mt0[
Aplicação se tornando cada vez mais concreta durante a atual pandemia do Sars-CoV-2 
]
<br>
<br>
<img src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1186%2Fs40246-021-00326-3/MediaObjects/40246_2021_326_Fig1_HTML.png">
]

---

# Porém ...

* Apesar de sua relevância, a implementação destas práticas em meios clínicos permanece um grande desafio

* De maneira geral, as análises dependem bastante de habilidades e expertise em bioinformática.

    + Seleção da plataforma de sequenciamento

    + Seleção e instalação das ferramentas e recursos
    
    + Preparação dos dados
    
    + Execução da análise
    
    + Interpretação dos resultados

.bg-washed-blue.b--dark-blue.ba.bw2.br3.shadow-5.ph4.mt0[
Estas questões apresentam grande barreira para a implementação de rotinas de sequenciamento em ambientes clínicos
]

???

* However the implementation of such practices in clinical settings are still a challenge because it heavily depends on bioinformatics skills and expertise, such as:

* Selecting the right sequencing platform
* Prepare the input data, analyse it and interpret the results in practical information and solutions

---

class: middle, center, inverse

# Urgência por soluções de programas completos e abrangentes

???

Therefore, there is a urgent need for a comprehensive solution that mitigates such hindrance.

---

# _Pipelines_ baseados em contêineres

.pull-left-bigger[
<img src="IMAGES/sc-software.svg" width="700px">
]

.pull-right-smaller[
<br>
* Para tentar mitigar estas barreiras, desenvolvemos 3 _pipelines_ sequenciais:

    + `r fontawesome("github")` [ngs-preprocess](https://github.com/fmalmeida/ngs-preprocess)
    + `r fontawesome("github")` [MpGAP](https://github.com/fmalmeida/mpgap)
    + `r fontawesome("github")` [bacannot](https://github.com/fmalmeida/bacannot)

* Adaptável a diferentes cenários
    + Illumina
    + Pacbio
    + ONT
]

???

In order to address this need, we developed three sequential pipelines that together (point to them in the image) provide a comprehensive workflow for raw data preprocessing, genome assembly and bacterial genome annotation.

All that adaptable to different scenarios, meaning that users can use different combinations of data from Illumina, Pacbio and Nanopore sequencing platforms.

---

# Implementação

.pull-left[
A implementação baseada em contêineres permite:

+ Portabilidade
    + Dependências empacotadas em `r fontawesome("docker")`, facilitando a instalação:
    + `docker pull <image>`
+ Reprodutibilidade
    + Nextflow assegura a execução do _pipeline_ para diferentes conjuntos de dados
+ Escalabilidade
    + Nextflow provê uma maneira simples de paralelizar e implementar trabalhos em qualquer máquina, de laptos a super computadores
+ Tolerância a erros
    + Nextflow rastreia cada processo e cria _checkpoints_ permitindo pontos de reinício em caso de erros
]

.pull-right[

**Distribuição**

* Pipelines disponíveis no github `r fontawesome("github")`
* Ferramentas empacotadas em imagens `r fontawesome("docker")` 

<img src="IMAGES/usage_overview_docker.png">

]

???

* The pipelines were implemented in Nextflow and all the tools are wrapped inside docker containers, which makes distribution and installation be quick and easy, only being required to download the pipeline from github and the docker images from dockerhub.

* Moreover, since the pipelines are implemented in nextflow, they are:

    + Extremely reproducible, meaning nextflow will ensure the same execution for different datasets
    
    + Highly scalable because nextflow provides easy parallelization of jobs and easy portability to clusters, meaning the pipelines can be easily used in any machine, from laptos to HPCs, from 1 to 1000s samples.

---

# Instalação e execução simplificadas

.pull-left[

* Os _pipelines_ são instalados e executados em comando único.

    + Parâmetros podem ser passados em linha de comando
    + Mas recomenda-se o uso de arquivos de configuração

```bash
nextflow run fmalmeida/ngs-preprocess -c config.txt
```

<img src="IMAGES/sc-execution-nf.png">

]

.pull-right[
* Arquivo de configuração

<br>
<img src="IMAGES/sc-config.png">
]

???

* Here we can see how simple it is to use the pipelines!

    + The command `nextflow run` will download the pipeline if not already available and execute it!

* All parameters can be set via the command line, however it is easier and cleaner to use the configuration file which is exemplified in the image on the right. This is the simplest way to configure and use the pipelines.

---

# Bacannot, _pipeline_ de anotação

.pull-left[
* Bacannot foi desenhado para ser bastante abrangente!

    + Mas modular
        + Análises podem ser "desligadas"
    
    + Escalonável de um a milhares de genomas
    
    + Relatórios claros e um aplicativo-web para exploração dos resultados
    
    + Muito bem documentado
        + https://bacannot.readthedocs.io/en/latest/

* Por ser escrito em módulos, pode rapidamente se adaptar a necessidades da comunidade:
    
    + Novos relatórios
    + Novas análises
    + Etc.
]


.pull-right[
<center>
<img src="IMAGES/sc-software-bacannot.png" width="450px">
</center>
]

???

* To exemplify one of the pipelines, I will discuss the implementation of the annotation pipeline.

* Bacannot is a comprehensive annotation pipeline which performs various types of annotations steps, from CDS and rRNA prediction, to focused annotation and detection of plasmids, genomic islands, prophages, integrative and conjugative elements, virulence and antimicrobial resistance genes, and more.

* Although comprehensive, the pipeline is modular, meaning that users can skip annotation steps if desired.

* Which also makes the it highly adaptable to the community needs, meaning that new steps and modules can be rapidly added because it does not require changes in the whole pipeline.

* All that while providing nice alternatives for results visualization and interpretation, such as: HTML reports, a genome browser, and a web app for a compact way of exploring and querying the results

---

# Bacannot _outputs_: Relatórios

* Example: Relatório AMR

<center>
<img src="IMAGES/amr-report-sc.png" width="95%">
</center>

???

* All that while providing nice alternatives for results visualization and interpretation, such as: HTML reports

* Here I show you the AMR report which have a tab summarizing the results of each specific tool.

---

# Bacannot _outputs_: Genome browser

* All annotation available in the browser

<center>
<img src="https://bacannot.readthedocs.io/en/latest/_images/jbrowse.png" width="90%">
</center>

???

* A genome browser which wraps up all the results obtained with the pipeline providing a beautiful way to visualize and investigate th results.

---

# Bacannot _outputs_: web (shiny) app

.pull-left-bigger[
![](IMAGES/bacannot_shiny.gif)
]

.pull-right-smaller[
* Indexação dos relatórios de anotação

* Visualização do navegador genômico

* Exploração e interrogação dos resultados armazenados em SQL

* BLAST dinâmico:

    + Maneira tabular para identificação das posições e interseção com anotação
    
    + Para visualização do alinhamento com [SequenceServer](https://sequenceserver.com/)
]

???

* And also a a web app for a compact way of exploring and querying the results.

* Here we can see different tabs which list all the HTML reports, a tab for querying the results stored in a SQLdb so users can rapidply filter and interrogate their data and a tab providing a way to BLAST your input and visualize whether the BLAST results intersect with the annotations results produced with the pipeline.

---

# Trabalhos realizados

* Multidrug-Resistant (MDR) _Klebsiella variicola_ Strains Isolated in a Brazilian Hospital Belong to New Clones
    
    + https://doi.org/10.3389/fmicb.2021.604031

<center>
```{r, echo=FALSE, out.width="45%"}
knitr::include_graphics(c("https://www.frontiersin.org/files/Articles/604031/fmicb-12-604031-HTML-r1/image_m/fmicb-12-604031-g002.jpg","https://www.frontiersin.org/files/Articles/604031/fmicb-12-604031-HTML-r1/image_m/fmicb-12-604031-g003.jpg"))
```
</center>

---

# Trabalhos realizados

* A novel multidrug resistant, non-Tn4401 genetic element (NTEKPC)-bearing, strain type of _Klebsiella pneumoniae_ isolated from an urban lake with drinking and recreational water reuse 
    
    + Em revisão (frontiers)

<center>
<img src="IMAGES/ggtree_test.svg" width="250"> <img src="IMAGES/gcluster.svg" width="400">
</center>

---

# Trabalhos realizados

* Genomic and physiological characterization of _Novosphingobium terrae_ sp. nov., an alphaproteobacterium isolated from Cerrado soil
    
    + Em revisão (Spectrum)

<center>
<img src="IMAGES/n_sp_dnaplotter.png" width="400"> <img src="IMAGES/Figure7_sulfur.svg" width="450">
</center>

---

# Perspectivas

* Criar um programa de **genômica comparativa** que receba os resultados do _pipeline_ de anotação

    + `bacannot-compare`
    + Pangenoma, filogenia, comparação de conteúdo funcional, resistência, STs, etc.

.pull-left[
<img src="IMAGES/ko_comparison_snapshot.png" width="400">
<img src="IMAGES/gingr_snapshot.png" width="400">
]

.pull-right[
<br>
<br>
<img src="IMAGES/pan_genome.svg">
<center>
```{r, echo=FALSE, out.width="30%"}
knitr::include_graphics(c("IMAGES/pangenome_frequency.svg", "IMAGES/pangenome_pie.svg", "IMAGES/pangenome_matrix.svg"))
```
</center>
]

---
class: inverse, middle, center

# Obrigado!
